app/src/main/java/http_server/ResponseBuilder.java

```diff
@@ -1,58 +1,101 @@
+  public ResponseBuilder(ResponseHandler requestHandler, RouteValidator routeValidator) {
     this.handler = requestHandler;
+    this.routeValidator = routeValidator;
   }
 
   public Response buildResponse(RequestReader request) {
     String path = request.getPath();
     String method = request.getMethod();
 
+    if (!routeValidator.hasPath(path)) {
+      return handler.buildUnexpectedResponse("404");
+    }
+
+    if (!routeValidator.hasMethod(path, method)) {
+      return handler.buildMethodOptionsResponse("allow", "GET, HEAD, OPTIONS", "405");
+    }
+
     if (path != null && method != null) {
       switch (path) {
-        case "/simple_get_with_body":
-          return handler.buildSimpleGetWithBodyResponse();
-
-        case "/simple_get":
-          if (method.equals("HEAD")) {
-            return handler.buildHeadResponse();
-          }
-          return handler.buildSimpleGetResponse();
-
-        case "/echo_body":
-          if (method.equals("POST")) {
-            String body = request.getBody();
-            if (body != null) {
-              return handler.buildEchoBodyResponse(body);
-            }
-          }
-          break;
-
-        case "/head_request":
-          if (method.equals("HEAD")) {
-            return handler.buildHeadResponse();
-          }
-          return handler.buildNonHeadResponse();
-
-        case "/method_options":
-          if (method.equals("OPTIONS")) {
-            return handler.buildMethodOptionsResponse("allow", "GET, HEAD, OPTIONS");
-          }
-
-        case "/method_options2":
-          if (method.equals("OPTIONS")) {
-            return handler.buildMethodOptionsResponse("allow", "GET, HEAD, OPTIONS, PUT, POST");
-          }
-
-        default:
-          break;
+        case "/simple_get_with_body" -> {
+          return buildSimpleGetWithBodyResponse(method);
+        }
+        case "/simple_get" -> {
+          return buildSimpleGetResponse(method);
+        }
+        case "/echo_body" -> {
+          return buildEchoBodyResponse(method, request.getBody());
+        }
+        case "/head_request" -> {
+          return buildHeadRequestResponse(method);
+        }
+        case "/method_options" -> {
+          return buildMethodOptionsResponse(method, "GET, HEAD, OPTIONS");
+        }
+        case "/method_options2" -> {
+          return buildMethodOptionsResponse(method, "GET, HEAD, OPTIONS, PUT, POST");
+        }
+        case "/redirect" -> {
+          return buildRedirectResponse(method);
+        }
+        default -> {
+        }
       }
     }
 
-    return handler.buildUnexpectedResponse();
+    return handler.buildUnexpectedResponse("405");
+  }
+
+  private Response buildSimpleGetWithBodyResponse(String method) {
+    if (method.equals("HEAD")) {
+      return handler.buildHeadResponse("200");
+    } else {
+      return handler.buildResponse("200", "Hello world");
+    }
+  }
+
+  private Response buildSimpleGetResponse(String method) {
```

Comment:
Another thought.. I'm wondering if these private methods here should actually start with "handle" instead since it's the one figuring out which `handler.buildXX())` method to use

Reply:
Ahh yes. that makes sense! I'll change the naming in these.